File: src/web/backend/app/schemas/knowledge.py
"""
知识 Pydantic 模式
定义疾病知识数据的请求和响应格式
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class DiseaseBase(BaseModel):
    """疾病基础模式"""
    name: str = Field(..., min_length=1, max_length=100, description="疾病名称")
    code: Optional[str] = Field(None, max_length=50, description="疾病编码")
    category: Optional[str] = Field(None, max_length=50, description="疾病分类")


class DiseaseCreate(DiseaseBase):
    """疾病创建请求模式"""
    symptoms: Optional[str] = Field(None, description="症状描述")
    causes: Optional[str] = Field(None, description="病因")
    treatments: Optional[str] = Field(None, description="治疗方法")
    prevention: Optional[str] = Field(None, description="预防措施")
    severity: Optional[float] = Field(0.0, ge=0, le=1, description="严重程度")


class DiseaseUpdate(BaseModel):
    """疾病更新请求模式"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="疾病名称")
    code: Optional[str] = Field(None, max_length=50, description="疾病编码")
    category: Optional[str] = Field(None, max_length=50, description="疾病分类")
    symptoms: Optional[str] = Field(None, description="症状描述")
    causes: Optional[str] = Field(None, description="病因")
    treatments: Optional[str] = Field(None, description="治疗方法")
    prevention: Optional[str] = Field(None, description="预防措施")
    severity: Optional[float] = Field(None, ge=0, le=1, description="严重程度")


class DiseaseResponse(DiseaseBase):
    """疾病响应模式"""
    id: int
    symptoms: Optional[str] = None
    causes: Optional[str] = None
    treatments: Optional[str] = None
    prevention: Optional[str] = None
    severity: Optional[float] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class DiseaseSearch(BaseModel):
    """疾病搜索请求模式"""
    keyword: Optional[str] = Field(None, description="搜索关键词")
    category: Optional[str] = Field(None, description="疾病分类")
    page: int = Field(1, ge=1, description="页码")
    page_size: int = Field(10, ge=1, le=100, description="每页数量")

File: src/web/frontend/src/api/knowledge.ts
/**
 * 知识库 API 调用函数
 * 负责调用后端知识库相关接口
 */
import { http } from '@/utils/request'

// API 基础路径
const API_BASE = '/knowledge'

/**
 * 病害知识类型
 */
export interface DiseaseKnowledge {
  id: number
  name: string
  code?: string
  category?: string
  symptoms?: string
  causes?: string
  treatments?: string
  prevention?: string
  severity?: number
  created_at: string
  updated_at: string
}

/**
 * 知识图谱节点类型
 */
export interface KnowledgeNode {
  id: string
  name: string
  type: string
  category?: string
  severity?: number
}

/**
 * 知识图谱关系类型
 */
export interface KnowledgeRelation {
  from: string
  to: string
  type: string
  properties?: Record<string, any>
}

/**
 * 获取病害知识列表
 * @param category 病害类别（可选）
 * @param skip 跳过记录数
 * @param limit 返回记录数
 * @returns 病害知识列表
 */
export async function getDiseaseKnowledge(
  category?: string,
  skip: number = 0,
  limit: number = 20
): Promise<DiseaseKnowledge[]> {
  const page = Math.floor(skip / limit) + 1
  return http.get<DiseaseKnowledge[]>(`${API_BASE}/search`, {
    params: { category, page, page_size: limit }
  })
}

/**
 * 获取病害知识详情
 * @param id 病害 ID
 * @returns 病害知识详情
 */
export async function getDiseaseDetail(id: number): Promise<DiseaseKnowledge> {
  return http.get<DiseaseKnowledge>(`${API_BASE}/${id}`)
}

/**
 * 搜索病害知识
 * @param keyword 搜索关键词
 * @param skip 跳过记录数
 * @param limit 返回记录数
 * @returns 搜索结果
 */
export async function searchDiseaseKnowledge(
  keyword: string,
  skip: number = 0,
  limit: number = 20
): Promise<DiseaseKnowledge[]> {
  const page = Math.floor(skip / limit) + 1
  return http.get<DiseaseKnowledge[]>(`${API_BASE}/search`, {
    params: { keyword, page, page_size: limit }
  })
}

/**
 * 获取知识图谱
 * @param diseaseId 病害 ID（可选）
 * @returns 知识图谱数据
 */
export async function getKnowledgeGraph(diseaseId?: number): Promise<{
  nodes: KnowledgeNode[]
  relations: KnowledgeRelation[]
}> {
  return http.get(`${API_BASE}/graph`, {
    params: diseaseId ? { disease_id: diseaseId } : undefined
  })
}

/**
 * 获取病害统计信息
 * @returns 统计数据
 */
export async function getDiseaseStats(): Promise<{
  total: number
  by_category: Record<string, number>
}> {
  return http.get(`${API_BASE}/stats`)
}
